rfcs/0000-nixpkgs-breaking-change-policy.md

```diff
@@ -0,0 +1,75 @@
+| breaking_changes | 2020-09-24 | Kevin Cox | (find a buddy later to help out with the RFC) | (names, to be nominated and accepted by RFC steering committee) | (name to be appointed by RFC steering committee) | (will contain links to implementation PRs) |
+
+# Summary
+
+This defines a policy for how to manage nixpkgs changes that break dependent derivations or tests. It specifies time frames and procedures so that maintainers of dependencies and dependents can know what to expect from each other when making breaking changes.
+
+This document is only focused on changes to derivation builds and tests it **does not** aim to make any opinion on changes that can break user machines or configurations.
+
+# Motivation
+
+Right now the procedure for making breaking changes varies widely by maintainer and is largely tribal knowledge. There are a wide variety of approaches ranging from:
+
+- After testing to confirm that a change is correct, merge to `master` and notify dependents that they need to update their packages.
+- Merge to `staging`, notify dependents then eventually merge to `master`.
+- Gather a large number of fixes in the PR branch then merge to `master` once most or all dependencies are fixed.
+
+This aims to provide a uniform approach so that everyone involved knows what to expect and to allow further workflows including automation to be build on top of these procedures.
+
+## Past Discussions
+
+https://discourse.nixos.org/t/please-fix-all-packages-which-were-broken-by-the-qt-wrapping-changes/6444
+
+# Detailed design
+
+## Goals
+
+The following procedure will be used for making breaking changes. It has a few primary goals:
+
+- Avoid putting more burden than necessary on the dependency maintainer. If the maintainers of core derivations face toil proportionally to the number of transitive dependencies they will quickly become overloaded. These maintainers are arguably the most critical to nixpkgs and  their load needs to be kept manageable.
+- Avoid unnecessarily breaking packages for any period of time. There are a number of users on the `*-unstable` channels and it is annoying if packages in their configuration get broken.
+- Avoid breaking the `master` branch as much as possible. The `master` branch is used by the `*-unstable` channels and breaking this branch means that no updates can be pushed. This includes critical security updates as well as regular feature updates.
+
+## Procedure
+
+1. The maintainer will prepare a PR with their intended changes.
+2. The maintainer should test a sample of dependent derivations to ensure that their change is not unnecessarily or unintentionally breaking. (Example: Ensure that a critical output file was not forgotten) Note that sometimes it **is** necessary to break all dependent packages, and the maintainer is not required to avoid this.
```

Comment:
Whether it is sometimes necessary to break all dependent packages is quite a claim and needs further motivation.

Reply:
I should probably remove the "all" but I'm sure there are cases where libraries make updates that change the build system or remove something. The point here is that the initial review should consider if the amount of breakage is reasonable. I will reword to make that more explicit.